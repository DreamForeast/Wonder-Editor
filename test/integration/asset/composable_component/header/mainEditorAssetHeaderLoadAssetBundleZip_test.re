open Wonder_jest;

/* open AssetTreeTwoLayerTypeTool; */

open Expect;

open Expect.Operators;

open Sinon;

/* open NodeAssetType; */

open Js.Promise;

open Js.Typed_array;

let _ =
  describe("MainEditorAssetHeader->load asset bundle zip", () => {
    let sandbox = getSandboxDefaultVal();

    beforeEach(() => {
      sandbox := createSandbox();

      MainEditorSceneTool.initState(~sandbox, ());

      MainEditorAssetHeaderLoadTool.prepareInspectorCanvas(sandbox);

      MainEditorSceneTool.prepareScene(sandbox);
    });

    afterEach(() => restoreSandbox(refJsObjToSandbox(sandbox^)));

    describe("test load asset bundle zip", () => {
      let _buildFakeZipData = [%bs.raw
        (wab, sab) => {|
  return [
    ["A.wab",{
      async: function(){
          return new Promise((resolve, _) => resolve(
new Uint8Array(wab)
          ))
      },
    }],
    ["SAB",{
      async: function(){
          return new Promise((resolve, _) => resolve(
          ))
      },
    }],
    ["SAB/B.sab",{
      async: function(){
          return new Promise((resolve, _) => resolve(
new Uint8Array(sab)
          ))
      },
    }]
  ]
|}
      ];

      let _getWABName = () => "A";

      let _getSABName = () => "B";

      let _prepare = () => {
        let addedFolderNodeId = MainEditorAssetIdTool.getNewAssetId();

        MainEditorAssetHeaderOperateNodeTool.addFolder();

        let wab = ArrayBuffer.make(3);
        let sab = ArrayBuffer.make(2);

        let obj =
          MainEditorAssetHeaderLoadZipTool.buildImportFakeJsZipCreateFunc(
            sandbox^,
            _buildFakeZipData(wab, sab),
          );

        MainEditorAssetTreeTool.Select.selectFolderNode(
          ~nodeId=addedFolderNodeId,
          (),
        );

        (obj, wab, sab);
      };

      beforeEach(() => {
        MainEditorAssetTool.buildFakeFileReader();
        LoadTool.buildFakeTextDecoder(LoadTool.convertUint8ArrayToBuffer);
        LoadTool.buildFakeTextEncoder();
        LoadTool.buildFakeURL(sandbox^);
        LoadTool.buildFakeLoadImage(.);

        EventListenerTool.buildFakeDom()
        |> EventListenerTool.stubGetElementByIdReturnFakeDom;
      });

      testPromise("should add to asset children", () => {
        let (obj, wab, sab) = _prepare();

        MainEditorAssetUploadTool.loadOneAssetBundleZip(
          ~sandbox,
          ~createJsZipFunc=() => obj,
          (),
        )
        |> then_(uploadedWDBNodeId =>
             BuildComponentTool.buildAssetChildrenNode()
             |> ReactTestTool.createSnapshotAndMatch
             |> resolve
           );
      });
      testPromise("should store assetBundle arrayBuffer", () => {
        let (obj, wab, sab) = _prepare();

        MainEditorAssetUploadTool.loadOneAssetBundleZip(
          ~sandbox,
          ~createJsZipFunc=() => obj,
          (),
        )
        |> then_(() =>
             (
               MainEditorAssetAssetBundleNodeTool.getAssetBundleNodeByName(
                 _getWABName(),
               )
               |> StateLogicService.getStateToGetData
               |> AssetBundleNodeAssetService.getAssetBundle,
               MainEditorAssetAssetBundleNodeTool.getAssetBundleNodeByName(
                 _getSABName(),
               )
               |> StateLogicService.getStateToGetData
               |> AssetBundleNodeAssetService.getAssetBundle,
             )
             |> expect == (wab, sab)
             |> resolve
           );
      });
    });
  });