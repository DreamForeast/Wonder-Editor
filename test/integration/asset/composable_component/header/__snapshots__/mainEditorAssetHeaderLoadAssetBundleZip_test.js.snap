// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainEditorAssetHeader->load asset bundle zip test load asset bundle zip should add to asset children 1`] = `
<article
  className="wonder-asset-assetChildren"
>
  <article
    className="wonder-asset-folderBox"
    id="folder-11"
    style={
      Object {
        "border": "0px",
      }
    }
  >
    <div
      className="item-ground"
      draggable={true}
      onDragEnd={[Function]}
      onDragEnter={[Function]}
      onDragLeave={[Function]}
      onDragOver={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
    />
    <div
      className="box-image"
    >
      <img
        src="./public/img/assetPackage.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        SAB
      </span>
    </div>
  </article>
  <article
    className="wonder-asset-fileBox "
    onClick={[Function]}
  >
    <div
      className="box-image"
    >
      <img
        onDragStart={[Function]}
        src="./public/img/assetBundle.png"
      />
    </div>
    <div
      className="item-text "
    >
      <span>
        A
      </span>
    </div>
  </article>
</article>
`;
